x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/indicator_match/threat_mapping/create_threat_signals.ts

```diff
@@ -169,14 +171,83 @@ export const createThreatSignals = async ({
     let list = await getDocumentList({ searchAfter: undefined });
     let documentCount = totalDocumentCount;
 
+    // this is re-assigned depending on max clause count errors
+    let chunkPage = itemsPerSearch;
+
     while (list.hits.hits.length !== 0) {
       verifyExecutionCanProceed();
-      const chunks = chunk(itemsPerSearch, list.hits.hits);
+      const chunks = chunk(chunkPage, list.hits.hits);
       ruleExecutionLogger.debug(`${chunks.length} concurrent indicator searches are starting.`);
       const concurrentSearchesPerformed =
         chunks.map<Promise<SearchAfterAndBulkCreateReturnType>>(createSignal);
+
       const searchesPerformed = await Promise.all(concurrentSearchesPerformed);
-      results = combineConcurrentResults(results, searchesPerformed);
+
+      // Did our searches fail with an error containing the maxClauseCount
+      // error message?
+
+      // TODO - delete this comment
+      // FOR PR TESTING PURPOSES ONLY
+      // uncomment this line..
+      // const maxClauseCountValue = MAX_CLAUSE_DEFAULT_VALUE;
+
+      // and comment out the below reduce function to force the
+      // max clause count error to re-appear.
+      const maxClauseCountValue = searchesPerformed.reduce<number>((acc, search) => {
+        const failureMessage: string | undefined = search.errors.find((err) =>
+          err.includes('failed to create query: maxClauseCount is set to')
+        );
+
+        const regex = /[0-9]/g;
+        const foundMaxClauseCountValue = failureMessage?.match(regex);
+
+        if (!isEmpty(foundMaxClauseCountValue) && foundMaxClauseCountValue != null) {
+          const tempVal = parseInt(foundMaxClauseCountValue.join(''), 10);
+          let val = MAX_CLAUSE_DEFAULT_VALUE;
+          if (tempVal === 0) {
+            // if the resulting max clause count error is 0, don't subtract 1 from that.
+            // rare case that really should not happen unless ES claims heap size is 0.
+            // ref: https://github.com/elastic/elasticsearch/issues/86136
+            val = tempVal;
+          } else {
+            // minus 1 since it needs to be one less than the max
+            // provided by elasticsearch error message.
+            val = tempVal - 1;
+          }
+          ruleExecutionLogger.warn(
+            `maxClauseCount error received from elasticsearch, setting IM rule page size to ${val}`
+          );
+          return val;
+        } else {
+          return acc;
+        }
+      }, MAX_CLAUSE_DEFAULT_VALUE);
+
+      if (maxClauseCountValue > MAX_CLAUSE_DEFAULT_VALUE) {
+        // parse the error message to acquire the number of maximum possible clauses
+        // allowed by elasticsearch. The sliced chunk is used in createSignal to generate
+        // threat filters.
+        chunkPage = maxClauseCountValue;
```

Comment:
I think correlation 1:1 with chunk size to max clause count would only work if rule has only one matching clause.

When I added another AND, rule started to fail

<img width="2547" alt="Screenshot 2024-04-03 at 13 23 51" src="https://github.com/elastic/kibana/assets/92328789/4e8048fa-c49a-4311-8420-4a928f4607c0">

Number of matching clauses should be accounted in determination of chunk page size